File: src/main/java/net/clgd/ccemux/emulation/EmulatedEnvironment.java
package net.clgd.ccemux.emulation;

import java.io.IOException;
import java.nio.file.Paths;
import java.util.zip.ZipInputStream;

import dan200.computercraft.ComputerCraft;
import dan200.computercraft.api.filesystem.IMount;
import dan200.computercraft.api.filesystem.IWritableMount;
import dan200.computercraft.core.computer.IComputerEnvironment;
import dan200.computercraft.core.filesystem.ComboMount;
import dan200.computercraft.core.filesystem.FileMount;
import dan200.computercraft.core.filesystem.JarMount;

public class EmulatedEnvironment implements IComputerEnvironment {

	private int nextID = 0;
	private final CCEmuX emu;
	
	EmulatedEnvironment(CCEmuX emu) {
		this.emu = emu;
	}
	
	@Override
	public int assignNewID() {
		return nextID++;
	}
	
	@Override
	public IMount createResourceMount(String domain, String subPath) {
		String path = Paths.get("assets", domain, subPath).toString().replace('\\', '/');
		
		if (path.startsWith("\\")) path = path.substring(1);
		try {
			return new ComboMount(new IMount[] {
				new JarMount(emu.ccJar, path),
				new CustomRomMount(new ZipInputStream(this.getClass().getResourceAsStream("/custom.rom")))
			});
		} catch (IOException e) {
			emu.logger.error("Failed to create resource mount", e);
			return null;
		}
	}
	
	@Override
	public IWritableMount createSaveDirMount(String path, long capacity) {
		return new FileMount(emu.dataDir.resolve(path).toFile(), capacity);
	}
	
	@Override
	public long getComputerSpaceLimit() {
		return Long.MAX_VALUE;
	}
	
	@Override
	public int getDay() {
		return (int)((emu.getTicksSinceStart() + 6000L) / 24000L) + 1;
	}
	
	@Override
	public double getTimeOfDay() {
		return ((emu.getTicksSinceStart() + 6000) % 24000) / 1000;
	}
	
	@Override
	public boolean isColour() {
		return true;
	}
	
	@Override
	public String getHostString() {
		return "ComputerCraft " + ComputerCraft.getVersion() + " (CCEmuX v" + CCEmuX.getVersion() + ")";
	}
}

File: src/main/kotlin/net/clgd/ccemux/emulation/EmulatedEnvironment.kt
package net.clgd.ccemux.emulation

import dan200.computercraft.ComputerCraft
import dan200.computercraft.core.computer.IComputerEnvironment
import dan200.computercraft.core.filesystem.ComboMount
import dan200.computercraft.core.filesystem.FileMount
import dan200.computercraft.core.filesystem.JarMount
import java.nio.file.Paths
import java.util.zip.ZipInputStream

class EmulatedEnvironment(val emu: CCEmuX) : IComputerEnvironment {
	private var nextID = 0

	override fun assignNewID() = nextID++

	override fun createResourceMount(domain: String, subPath: String): ComboMount {
		var path = Paths.get("assets", domain, subPath).toString().replace('\\', '/')

		if (path.startsWith('\\')) {
			path = path.substring(1)
		}

		return ComboMount(
			arrayOf(
				JarMount(emu.ccJar, path),
				CustomRomMount(
					ZipInputStream(EmulatedEnvironment::class.java.getResourceAsStream("/custom.rom"))
				)
			)
		)
	}

	override fun createSaveDirMount(path: String, capacity: Long): FileMount {
		return FileMount(emu.dataDir.resolve(path).toFile(), capacity)
	}

	override fun getComputerSpaceLimit() = Long.MAX_VALUE

	override fun getDay() = ((emu.getTicksSinceStart() + 6000L) / 24000L + 1).toInt()

	override fun getHostString() = "ComputerCraft %s (CCEmuX v%s)".format(ComputerCraft.getVersion(), CCEmuX.getVersion())

	override fun getTimeOfDay() = ((emu.getTicksSinceStart() + 6000) % 24000) / 1000.0

	override fun isColour() = true
}
